File: src/Bullet.java
import processing.core.PApplet;

/** Simple projectile used by the Minigun weapon. */
public class Bullet {
    float x, y;
    float speedX, speedY;
    float damage;
    float size = 8;

    public Bullet(float x, float y, float speedX, float speedY, float damage) {
        this.x = x;
        this.y = y;
        this.speedX = speedX;
        this.speedY = speedY;
        this.damage = damage;
    }

    public void update() {
        x += speedX;
        y += speedY;
    }

    public boolean isOffScreen(int width, int height) {
        return x < -20 || x > width + 20 || y < -20 || y > height + 20;
    }

    public boolean checkHit(Ball b) {
        return PApplet.dist(x, y, b.x, b.y) < GameConfig.RADIUS + size / 2;
    }

    public void draw(PApplet p) {
        p.pushMatrix();
        p.translate(x, y, 0);
        p.noStroke();
        p.fill(255, 220, 40);
        p.sphere(size / 2f);
        p.popMatrix();
    }
}

File: Bullet.js
/* global p5, Ball, GameConfig */

class Bullet {
    constructor(x, y, speedX, speedY, damage) {
        this.x = x;
        this.y = y;
        this.speedX = speedX;
        this.speedY = speedY;
        this.damage = damage;
        this.size = 8;
    }

    update() {
        this.x += this.speedX;
        this.y += this.speedY;
    }

    isOffScreen(width, height) {
        return this.x < -20 || this.x > width + 20 || this.y < -20 || this.y > height + 20;
    }

    checkHit(b) {
        return dist(this.x, this.y, b.x, b.y) < GameConfig.RADIUS + this.size / 2;
    }

    draw() {
        push();
        translate(this.x, this.y, 0);
        noStroke();
        fill(255, 220, 40);
        sphere(this.size / 2);
        pop();
    }
}